File: Nordeus-Challenge/Assets/_Scripts/Models/Player/PlayerStateResponse.cs
using System;
using System.Collections.Generic;

[Serializable]
public class PlayerStateResponse
{
    public string id;
    public string characterName;
    public int    level;
    public int    currentXp;
    public int    xpToNextLevel;         // 2^(level-1): 1, 2, 4, 8, 16 …
    public int    gold;
    public int    currentHp;             // 0 = full HP (fresh player)
    public int    currentMana;
    public int    currentStamina;
    public CharacterStats   stats;
    public List<string>     knownMoves;    // all moves ever learned
    public List<string>     equippedMoves; // the 4 moves for the next combat
    public int              pendingStatPoints;
    public List<string>     inventory;     // item IDs in the player's bag
    public Equipment        equipment;     // equipped items by slot

    public float XpProgress      => xpToNextLevel > 0 ? (float)currentXp / xpToNextLevel : 0f;
    public bool  HasPendingLevelUp => pendingStatPoints > 0;
}

File: backend/src/main/java/com/example/backend/dtos/responces/PlayerStateResponse.java
package com.example.backend.dtos.responces;

import com.example.backend.models.characters.CharacterStats;
import com.example.backend.models.items.Equipment;
import com.example.backend.models.players.PlayerState;

import java.util.List;
import java.util.UUID;

public record PlayerStateResponse(
        UUID id,
        String characterName,
        int level,
        int currentXp,
        int xpToNextLevel,
        CharacterStats stats,
        List<String> knownMoves,
        List<String> equippedMoves,
        int pendingStatPoints,
        List<String> inventory,
        Equipment equipment,
        int gold,
        int currentHp,
        int currentMana,
        int currentStamina,
        int maxHp,
        int maxMana,
        int maxStamina
) {
    public static PlayerStateResponse from(PlayerState ps) {
        return new PlayerStateResponse(
                ps.getId(),
                ps.getCharacterName(),
                ps.getLevel(),
                ps.getCurrentXp(),
                ps.getXpToNextLevel(),
                ps.getStats(),
                ps.getKnownMoves(),
                ps.getEquippedMoves(),
                ps.getPendingStatPoints(),
                ps.getInventory(),
                ps.getEquipment(),
                ps.getGold(),
                ps.getCurrentHp(),
                ps.getCurrentMana(),
                ps.getCurrentStamina(),
                0, 0, 0
        );
    }

    public static PlayerStateResponse from(PlayerState ps, int maxHp, int maxMana, int maxStamina) {
        return new PlayerStateResponse(
                ps.getId(),
                ps.getCharacterName(),
                ps.getLevel(),
                ps.getCurrentXp(),
                ps.getXpToNextLevel(),
                ps.getStats(),
                ps.getKnownMoves(),
                ps.getEquippedMoves(),
                ps.getPendingStatPoints(),
                ps.getInventory(),
                ps.getEquipment(),
                ps.getGold(),
                ps.getCurrentHp(),
                ps.getCurrentMana(),
                ps.getCurrentStamina(),
                maxHp, maxMana, maxStamina
        );
    }
}
